### PR TITLE
Add bag.filter operator

### PR DESCRIPTION
**Pull request by mudathirmahgoub**

### CODE REVIEW

File: src/theory/bags/bags_utils.cpp [hunk 1]
```diff
@@ -683,11 +708,42 @@ Node NormalForm::evaluateBagMap(TNode n)
...
+  Node ret = BagsUtils::constructConstantBagFromElements(t, mappedElements);
+  return ret;
+}
+
+Node BagsUtils::evaluateBagFilter(TNode n)
+{
+  Assert(n.getKind() == BAG_FILTER);
+
+  // (bag.filter p (as bag.empty (Bag T)) = (as bag.empty (Bag T))
+  // (bag.filter p (bag "a" 3) ((bag "b" 2))) =
```
Comment (ajreynol) [inline src/theory/bags/bags_utils.cpp:720]: This line doesnt parse, do you mean a disjoint union here?

**Review by ajreynol — CHANGES_REQUESTED** [review]
It looks like the inferences involve only positive membership, is there something we have to do for negative membership into filter?

File: src/theory/bags/bags_utils.cpp (hunk 1 above)
Comment (mudathirmahgoub) [reply src/theory/bags/bags_utils.cpp]: Yes, fixed.

**Comment by mudathirmahgoub** [comment]
> It looks like the inferences involve only positive membership, is there something we have to do for negative membership into filter?

I guess not. Negative membership here means multiplicity is zero. 
From this table, I don't think we need inference rules for completeness. 
|(p x)| (bag.member x A) |(bag.member x (filter p A)) | Comment|
| --- | ----------- |------------|------------|
| true | 0 | 0 | - |
| true | 0 | 1 | unsat (down rule) |
| true | 1 | 0 | unsat (up rule) |
| true | 1 | 1 | sat (down + up) |
| false | 0 | 0 | - |
| false | 0 | 1 | unsat (down rule) |
| false | 1 | 0 | sat (up rule) |
| false | 1 | 1 | unsat (up rule) |

**Review by ajreynol — APPROVED** [review]

**Pull request merged**